segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/MetadataStore.java

```diff
@@ -0,0 +1,717 @@
+                                 Exceptions.unwrap(ex).getMessage());
+
+                         CompletableFuture<T> f = this.connector.getMetadataCleanup().get().thenComposeAsync(v -> toTry.get(), this.executor);
+                         f.thenAccept(result::complete);
+                         Futures.exceptionListener(f, result::completeExceptionally);
+                     } else {
+                         result.completeExceptionally(ex);
+                     }
+                 } catch (Throwable t) {
+                     result.completeExceptionally(t);
+                     throw t;
+                 }
+
+                 return null;
+             });
+
+        return result;
+    }
+
+    //endregion
+
+    //region Helper Classes
+
+    /**
+     * Mediates communication between a MetadataStore instance and an external caller.
+     */
+    @RequiredArgsConstructor
+    @Getter
+    public static class Connector {
+        /**
+         * The {@link ContainerMetadata} to bind to. All assignments are vetted from here, but the Metadata is not touched
+         * directly from this component.
+         */
+        @NonNull
+        private final ContainerMetadata containerMetadata;
+
+        /**
+         * A Function that, when invoked, updates the metadata with a new Segment Id assignment.
+         */
+        @NonNull
+        private final MapSegmentId mapSegmentId;
+
+        /**
+         * A Function that, when invoked, immediately deletes a Segment. This is invoked when it is certain that the Segment
+         * is not used at all, either because it is not loaded in the Metadata or because its Metadata Entry is corrupt.
+         */
+        @NonNull
+        private final DirectDeleteSegment directDeleteSegment;
+
+        /**
+         * A Function that, when invoked, submits a Segment for eventual deletion. This will be invoked when it is possible
+         * that the Segment is still actively being used, and we need a graceful deletion.
+         */
+        @NonNull
+        private final LazyDeleteSegment lazyDeleteSegment;
+
+        /**
+         * A Supplier that, when invoked, executes the Segment Container's Metadata Cleanup task which is responsible with
+         * evicting inactive Segments. This is invoked when a Segment Id assignment is in progress which cannot complete
+         * due to {@link TooManyActiveSegmentsException} being thrown.
+         */
+        @NonNull
+        private Supplier<CompletableFuture<Void>> metadataCleanup;
+
+        @FunctionalInterface
+        public interface MapSegmentId {
+            CompletableFuture<Long> apply(long segmentId, SegmentProperties segmentProperties, Duration timeout);
+        }
+
+        @FunctionalInterface
+        public interface DirectDeleteSegment {
+            CompletableFuture<Void> apply(String segmentName, Duration timeout);
```

Comment:
Why is the parameter here segment name while for `LazyDeleteSegment` it is the segment id? I'm wondering why it is not uniform.

Reply:
LazyDelete is used for Segments which are active in the metadata - we need to submit a DeleteSegmentOperation to properly delete it.

DirectDelete is used for Segments which are not active in the metadata - we directly delete from Tier2. Otherwise we'd have to needlessly invoke `getOrAssignId` just so that we can delete it.